### ISSUE TITLE
Type parameter for a PrintfFormat handling function defined in another file prevents compilation

### ISSUE THREAD

**Issue by rbauduin** [dotnet/fsharp#8434]
**Summary**
A function with a type argument and handling an argument of type `PrintfFormat` must be defined in the same file as its call sites. The same function without type argument can be placed in a different file without problem.

**Repro steps**

Here is an archive to reproduce the problem (unzip then `dotnet restore`and `dotnet run`): [fsharp-printf-typing-problem-master.zip](https://github.com/dotnet/fsharp/files/4139973/fsharp-printf-typing-problem-master.zip). Below is a description detailing the steps reproducing the problem.

Define a function handling a `PrintfFormat` argument in file `Utils.fs`
```
module Utils
    open System
    let formatStringHandler  (callback: string -> Type list-> obj list -> 'ReturnType)
                             (format:PrintfFormat<'PrintfType,_,_,'ReturnType>)
                             :'PrintfType =
        sprintf format
```

Now define a function `callFromOtherFile<'a>` in another module `DBA` in file `DBA.fs` calling this `PrintfFormat` handling function, 
```
let callback (format:string) (types:System.Type list) (values:obj list) =
    "result"
let callFromOtherFile<'a> format =
    Utils.formatStringHandler callback format
```
The type argument is not used in this example code, but is required in real usage code I want to compile (`format` is a parameterised SQL query, and `'a` is the single case union type in which to wrap the returned value). The type argument is causing the problem, hence its presence here.

Call that function from `Program.fs`:
```
let s2 = DBA.callFromOtherFile<string> "insert into instances(entity_id) VALUES (%d,%d) RETURNING id" 4 5
```
It causes a compilation error `error FS0003: This value is not a function and cannot be applied.`

**Expected behavior**

I would expect the code to compile.

**Actual behavior**

An compilation error occurs: `error FS0003: This value is not a function and cannot be applied.`

**Known workarounds**

The code works fine if the function (renamed to `callFromSameFile<'a>`) is defined in the same file as its call (`Program.fs`):
```
let callFromSameFile<'a> format =
    Utils.formatStringHandler callback format
```
This works, but introduces code duplication as the function cannot be placed in a module and reused in several places.

Removing the type parameter from `callFromOtherFile`and its call also compiles and runs fine.
So this compiles fine
```
let callFromSameFile format =
    Utils.formatStringHandler callback format
```
while this doesn't:
```
let callFromSameFile<'a> format =
    Utils.formatStringHandler callback format
```
But this is not really a workaround as it doesn't have the type argument.

**Related information**

I don't think this is relevant, but here is some info
* Operating system: Ubuntu 18.04
* .NET Runtime kind : Dotnet Core 3.1.100

**Comment by dsyme** [comment]
This is by-design. 

1. In the first file I see the warning:

```
Warning	FS0064	This construct causes code to be less generic than indicated by the type annotations. The type variable 'ReturnType has been constrained to be type 'string'.	t5	C:\GitHub\dsyme\tmp\t5\File1.fs	6	Active
```
for code
```
module Utils
open System
let formatStringHandler  (callback: string -> Type list-> obj list -> 'ReturnType)
                         (format:PrintfFormat<'PrintfType,_,_,'ReturnType>)
                         :'PrintfType =
    sprintf format

```


2. When you do this:

```
let callFromOtherFile<'a> format =
    Utils.formatStringHandler callback format
```

you are preventing the funciton `callFromOtherFile` from being fully generic (allowing it to have only one type parameter, unrelated to its implementation)

**Comment by rbauduin** [comment]
Thanks for having a look at it. I think that in simplifying the example code I ended up providing code with a warning invalidating my report. However, here is code ([fsharp-printf-typing-problem-master.zip](https://github.com/dotnet/fsharp/files/5174283/fsharp-printf-typing-problem-master.zip))  without warning showing the same behavior.

Could you have another look?

**Comment by dsyme** [comment]
The problem is this:

```
let callFromOtherFile<'a> format =
    Utils.formatStringHandler callback format
```

As mentioned above you are preventing the function callFromOtherFile from being fully generic (allowing it to have only one type parameter, unrelated to its implementation)

Changing to 

```
let callFromOtherFile format =
    Utils.formatStringHandler callback format
```

means automatic generalization happens and you have what you need.

**Comment by rbauduin** [comment]
Thanks a lot for indicating the needed change, I got it working!

In case anyone gets the same problem in the future and reads this, here is some more info.
I thought I required the type argument in my application code, where I have code like this:
```
let myFunction<'a> myArg = 
   ....
   ..typeof<'a>...
   ...
```
However, as the type signature of that function (as reported by ionide) is `int -> DBAction<'a>` I could remove the type argument and keep all code working by explicitly specifying the return type:
```
let myFunction myArg : DBAction<'a>= 
   ...
```
Tye type inference will occur for the return type, determining `'a`, making it usable in the function code.
